designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,308 @@
+
+## What can be Debugged Locally
+
+### <a id="sam-template-resource-local"></a> SAM Template Resources (Local Invoke)
+
+SAM Template Resources of type `AWS::Serverless::Function` represent Lambda functions. Lambda function code referenced by these resources can be locally Run or Debugged. The Toolkit uses SAM CLI to invoke the Lambda function, emulating how the function is run on AWS. A debugger can be attached to the invoked Lambda function code, and the event passed into the Lambda function can be customized.
+
+### SAM Template Resources (API Gateway style Local Invoke)
+
+SAM Template Resources that contain an event of type [Api](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-property-function-api.html), can be hosted in a web server for local development and iteration. A REST request triggers the Lambda handler as if it were called by API Gateway, and the Lambda function receives an API Gateway based event.
+
+### Standalone Lambda Function Handlers
+
+Lambda Function Handler code can be locally Run or Debugged, even if it does not belong to a SAM Application. The Toolkit produces a temporary SAM Application to contain the handler code. This temporary SAM Application is handled as mentioned [earlier](#sam-template-resource-local). At the end of the debug session, the temporary SAM Application is removed.
+
+In this mode, any SAM Templates that reference a Handler are ignored. This prevents confusion/errors introduced when trying to perform a reverse-lookup between code and SAM Template Resources (examples include incorrectly determining a function's lambda handler string, or situations where more than one resource references the same function).
+
+The Toolkit does not provide support for locally running or debugging standalone Lambda function handlers as API Gateway calls. The code should be referenced from a SAM Template in order to use the API Gateway style debugging features mentioned in the earlier section.
+
+## <a id="debug-config"></a> What can be configured for a Debug session?
+
+The following parameters influence a debug session. These are user-configured, and are referenced by the various [debugging experiences](#debugging-experiences).
+
+| Property                | Description                                          | Used by Standalone Lambda Handler | Used by SAM Template Resources |
+| ----------------------- | ---------------------------------------------------- | --------------------------------- | ------------------------------ |
+| SAM Template            | Path to SAM Template file                            |                                   | x                              |
+| SAM Template Resource   | Name of resource within SAM Template                 |                                   | x                              |
```

Comment:
What type of resource does this represent? Can I pick and `S3::Bucket`?

Reply:
Revised to "Name of lambda function-based resource within SAM Template" -- debug validation would have to inform users if the specified resource is not a supported type.

I will add an appendix for known validations.